src/chessboard/index.tsx

```diff
@@ -25,53 +26,104 @@ export type ClearPremoves = {
+    const [clientWindow, setClientWindow] = useState<Window>();
+    const [backendSet, setBackendSet] = useState(false);
+    const [isMobile, setIsMobile] = useState(false);
+    const [boardWidth, setBoardWidth] = useState(props.boardWidth);
 
-  const boardRef = useRef<HTMLObjectElement>(null);
+    const boardRef = useRef<HTMLObjectElement>(null);
 
-  useEffect(() => {
-    setIsMobile("ontouchstart" in window);
-    setBackendSet(true);
-    setClientWindow(window);
-  }, []);
+    useEffect(() => {
+      setIsMobile("ontouchstart" in window);
+      setBackendSet(true);
+      setClientWindow(window);
+    }, []);
 
-  useEffect(() => {
-    if (props.boardWidth === undefined && boardRef.current?.offsetWidth) {
-      const resizeObserver = new ResizeObserver(() => {
-        setBoardWidth(boardRef.current?.offsetWidth as number);
-      });
-      resizeObserver.observe(boardRef.current);
+    useEffect(() => {
+      if (props.boardWidth === undefined && boardRef.current?.offsetWidth) {
+        const resizeObserver = new ResizeObserver(() => {
+          setBoardWidth(boardRef.current?.offsetWidth as number);
+        });
+        resizeObserver.observe(boardRef.current);
 
-      return () => {
-        resizeObserver.disconnect();
-      };
-    }
-  }, [boardRef.current, clientWindow]);
+        return () => {
+          resizeObserver.disconnect();
+        };
+      }
+    }, [boardRef.current, clientWindow]);
 
-  const backend = customDndBackend || (isMobile ? TouchBackend : HTML5Backend);
+    const backend =
+      customDndBackend || (isMobile ? TouchBackend : HTML5Backend);
 
-  return backendSet && clientWindow ? (
-    <ErrorBoundary>
-      <div style={{ display: "flex", flexDirection: "column", width: "100%" }}>
-        <div ref={boardRef} style={{ width: "100%" }} />
-        <DndProvider
-          backend={backend}
-          context={clientWindow}
-          options={customDndBackend ? customDndBackendOptions : undefined}
+    return backendSet && clientWindow ? (
+      <ErrorBoundary>
+        <div
+          style={{ display: "flex", flexDirection: "column", width: "100%" }}
         >
-          {boardWidth && (
-            <ChessboardProvider boardWidth={boardWidth} {...otherProps} ref={ref}>
-              <CustomDragLayer />
-              <Board />
-            </ChessboardProvider>
-          )}
-        </DndProvider>
-      </div>
-    </ErrorBoundary>
-  ) : null;
-});
+          <div ref={boardRef} style={{ width: "100%" }} />
+          <DndProvider
+            backend={backend}
+            context={clientWindow}
+            options={customDndBackend ? customDndBackendOptions : undefined}
+          >
+            {boardWidth && (
+              <ChessboardProvider
+                boardWidth={boardWidth}
+                {...otherProps}
+                ref={ref}
+              >
+                <CustomDragLayer />
+                {/* TODO manage spare piece components in a normal way */}
```

Comment:
This is the most difficult part now. Spare pieces should be inside the `ReactDnDProvide`, I need some time for thinking and understanding how to make them maximum customizable for devs, who uses this library